lib/datadog/di/transport.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+require_relative 'error'
+
+module Datadog
+  module DI
+    # Transport for sending probe statuses and snapshots to local agent.
+    #
+    # Handles encoding of the payloads into multipart posts if necessary,
+    # body formatting/encoding, setting correct headers, etc.
+    #
+    # The transport does not handle batching of statuses or snapshots -
+    # the batching should be implemented upstream of this class.
+    #
+    # Timeout settings are forwarded from agent settings to the Net adapter.
+    #
+    # The send_* methods raise Error::AgentCommunicationError on errors
+    # (network errors and HTTP protocol errors). It is the responsibility
+    # of upstream code to rescue these exceptions appropriately to prevent them
+    # from being propagated to the application.
+    #
+    # @api private
+    class Transport
+      DIAGNOSTICS_PATH = '/debugger/v1/diagnostics'
+      INPUT_PATH = '/debugger/v1/input'
+
+      def initialize(agent_settings)
+        # Note that this uses host, port, timeout and TLS flag from
+        # agent settings.
+        @client = Datadog::Core::Transport::HTTP::Adapters::Net.new(agent_settings)
+      end
+
+      def send_diagnostics(payload)
+        event_payload = Datadog::Core::Vendor::Multipart::Post::UploadIO.new(
+          StringIO.new(JSON.dump(payload)), 'application/json', 'event.json'
+        )
+        payload = {'event' => event_payload}
+        send_request('Probe status submission', DIAGNOSTICS_PATH, payload)
+      end
+
+      def send_input(payload)
+        send_request('Probe snapshot submission', INPUT_PATH, payload,
+          headers: {'content-type' => 'application/json'},)
+      end
+
+      private
+
+      attr_reader :client
+
+      def send_request(desc, path, payload, headers: {})
+        # steep:ignore:start
+        env = OpenStruct.new(
+          path: path,
+          form: payload,
+          headers: headers,
+        )
+        # steep:ignore:end
+        response = client.post(env)
+        unless (200..299).cover?(response.code)
```

Comment:
There is a method in [Response class](https://github.com/DataDog/dd-trace-rb/blob/f0dd28e5640dfa3b3c8666fa5d6910651eaa7478/lib/datadog/core/transport/http/adapters/net.rb#L120)

```suggestion
        unless response.ok?
```